File: jdroid-java-webapp-sample/src/main/java/com/jdroid/javaweb/sample/api/controller/FirestoreController.java
package com.jdroid.javaweb.sample.api.controller;

import com.google.cloud.firestore.GeoPoint;
import com.jdroid.java.collections.Lists;
import com.jdroid.java.collections.Maps;
import com.jdroid.java.date.DateUtils;
import com.jdroid.java.http.MimeType;
import com.jdroid.java.utils.RandomUtils;
import com.jdroid.javaweb.api.AbstractController;
import com.jdroid.javaweb.sample.firebase.SampleEntity;
import com.jdroid.javaweb.sample.firebase.SampleFirestoreRepository;
import com.jdroid.javaweb.sample.firebase.SampleInnerEntity;

import org.springframework.stereotype.Controller;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestMethod;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.ResponseBody;

import java.util.List;
import java.util.Map;

@Controller
@RequestMapping("/firestore")
public class FirestoreController extends AbstractController {

	private SampleFirestoreRepository repository = new SampleFirestoreRepository();

	private static String lastId;
	
	@RequestMapping(value = "/getAll", method = RequestMethod.GET, produces = MimeType.JSON_UTF8)
	@ResponseBody
	public String getAll() {
		return autoMarshall(repository.getAll());
	}
	
	@RequestMapping(value = "/get", method = RequestMethod.GET, produces = MimeType.JSON_UTF8)
	@ResponseBody
	public String get(@RequestParam String id) {
		return autoMarshall(repository.get(id));
	}
	
	@RequestMapping(value = "/getByField", method = RequestMethod.GET, produces = MimeType.JSON_UTF8)
	@ResponseBody
	public String getByField(@RequestParam String field, @RequestParam String value) {
		return autoMarshall(repository.getByField(field, value));
	}
	
	@RequestMapping(value = "/getByFieldMultipleValues", method = RequestMethod.GET, produces = MimeType.JSON_UTF8)
	@ResponseBody
	public String getByField(@RequestParam String field, @RequestParam String value1, @RequestParam String value2) {
		return autoMarshall(repository.getByField(field, value1, value2));
	}

	@RequestMapping(value = "/add", method = RequestMethod.GET, produces = MimeType.JSON_UTF8)
	@ResponseBody
	public String add(@RequestParam(required = false, defaultValue = "false") Boolean autoId, @RequestParam(required = false, defaultValue = "false") Boolean withSubcollection) {
		SampleEntity entity = new SampleEntity();
		if (!autoId) {
			entity.setId(RandomUtils.getLong().toString());
		}
		entity.setStringField(RandomUtils.getLong().toString());
		entity.setLongField(RandomUtils.getLong());
		entity.setFloatField(1.12345F);
		entity.setTimestamp(DateUtils.now());
		Map<String, String> stringMap = Maps.newHashMap();
		stringMap.put("key1", "value1");
		stringMap.put("key2", "value2");
		entity.setStringMap(stringMap);
		Map<String, Object> objectMap = Maps.newHashMap();
		objectMap.put("key1", "value1");
		objectMap.put("key2", 2);
		objectMap.put("key3", new GeoPoint(-12.323434D, 34.34534543D));
		entity.setObjectMap(objectMap);
		entity.setGeoPoint(new GeoPoint(-12.323434D, 34.34534543D));
		entity.setStringArray(Lists.newArrayList("a", "b"));
		if (withSubcollection) {
			entity.setSubCollection(Lists.newArrayList(new SampleEntity(), new SampleEntity()));
		}
		
		SampleInnerEntity sampleInnerEntity = new SampleInnerEntity();
		sampleInnerEntity.setStringField(RandomUtils.getLong().toString());
		sampleInnerEntity.setLongField(RandomUtils.getLong());
		
		entity.setComposite(sampleInnerEntity);
		
		repository.add(entity);
		lastId = entity.getId();
		
		return autoMarshall(entity);
	}
	
	@RequestMapping(value = "/addAll", method = RequestMethod.GET, produces = MimeType.JSON_UTF8)
	@ResponseBody
	public String addAll(@RequestParam(required = false, defaultValue = "false") Boolean autoId) {
		
		List<SampleEntity> items = Lists.newArrayList();
		SampleEntity entity = new SampleEntity();
		if (!autoId) {
			entity.setId(RandomUtils.getLong().toString());
		}
		entity.setStringField(RandomUtils.getLong().toString());
		items.add(entity);
		entity = new SampleEntity();
		if (!autoId) {
			entity.setId(RandomUtils.getLong().toString());
		}
		entity.setStringField(RandomUtils.getLong().toString());
		items.add(entity);
		
		repository.addAll(items);
		
		return autoMarshall(items);
	}
	
	@RequestMapping(value = "/update", method = RequestMethod.GET, produces = MimeType.JSON_UTF8)
	@ResponseBody
	public String update() {
		SampleEntity entity = new SampleEntity();
		entity.setId(lastId);
		entity.setStringField(RandomUtils.getLong().toString());
		repository.update(entity);
		return autoMarshall(entity);
	}

	@RequestMapping(value = "/remove", method = RequestMethod.GET)
	public void remove(@RequestParam String id) {
		repository.remove(id);
	}

	@RequestMapping(value = "/removeAll", method = RequestMethod.GET)
	public void removeAll() {
		repository.removeAll();
		lastId = null;
	}
}

File: jdroid-java-webapp-sample/src/main/java/com/jdroid/javaweb/sample/api/controller/FirestoreController.kt
package com.jdroid.javaweb.sample.api.controller

import com.google.cloud.firestore.GeoPoint
import com.jdroid.java.collections.Lists
import com.jdroid.java.collections.Maps
import com.jdroid.java.date.DateUtils
import com.jdroid.java.http.MimeType
import com.jdroid.java.utils.RandomUtils
import com.jdroid.java.utils.TypeUtils
import com.jdroid.javaweb.api.AbstractController
import com.jdroid.javaweb.sample.firebase.SampleEntity
import com.jdroid.javaweb.sample.firebase.SampleFirestoreRepository
import com.jdroid.javaweb.sample.firebase.SampleInnerEntity

import org.springframework.stereotype.Controller
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RequestMethod
import org.springframework.web.bind.annotation.RequestParam
import org.springframework.web.bind.annotation.ResponseBody

@Controller
@RequestMapping("/firestore")
class FirestoreController : AbstractController() {

    companion object {

        private var lastId: String? = null
    }

    private val repository = SampleFirestoreRepository()

    @RequestMapping(value = ["/getAll"], method = [RequestMethod.GET], produces = [MimeType.JSON_UTF8])
    @ResponseBody
    fun getAll(@RequestParam(required = false) orderByField: String, @RequestParam(required = false) limit: String): String {
        return autoMarshall(object : SampleFirestoreRepository() {
            override fun getOrderByField(): String {
                return orderByField
            }

            override fun getLimit(): Int? {
                return TypeUtils.getInteger(limit)
            }
        }.all)
    }

    @RequestMapping(value = ["/get"], method = [RequestMethod.GET], produces = [MimeType.JSON_UTF8])
    @ResponseBody
    operator fun get(@RequestParam id: String): String {
        return autoMarshall(repository.get(id))
    }

    @RequestMapping(value = ["/getByField"], method = [RequestMethod.GET], produces = [MimeType.JSON_UTF8])
    @ResponseBody
    fun getByField(@RequestParam field: String, @RequestParam value: String): String {
        return autoMarshall(repository.getByField(field, value))
    }

    @RequestMapping(value = ["/getByFieldMultipleValues"], method = [RequestMethod.GET], produces = [MimeType.JSON_UTF8])
    @ResponseBody
    fun getByField(@RequestParam field: String, @RequestParam value1: String, @RequestParam value2: String): String {
        return autoMarshall(repository.getByField(field, value1, value2))
    }

    @RequestMapping(value = ["/getByIds"], method = [RequestMethod.GET], produces = [MimeType.JSON_UTF8])
    @ResponseBody
    fun getByIds(@RequestParam value1: String, @RequestParam value2: String): String {
        return autoMarshall(repository.getByIds(Lists.newArrayList(value1, value2)))
    }

    @RequestMapping(value = ["/add"], method = [RequestMethod.GET], produces = [MimeType.JSON_UTF8])
    @ResponseBody
    fun add(@RequestParam(required = false, defaultValue = "false") autoId: Boolean, @RequestParam(required = false, defaultValue = "false") withSubcollection: Boolean): String {
        val entity = SampleEntity()
        if (!autoId) {
            entity.id = RandomUtils.getLong().toString()
        }
        entity.stringField = RandomUtils.getLong().toString()
        entity.longField = RandomUtils.getLong()
        entity.floatField = 1.12345f
        entity.timestamp = DateUtils.now()
        val stringMap = Maps.newHashMap<String, String>()
        stringMap["key1"] = "value1"
        stringMap["key2"] = "value2"
        entity.stringMap = stringMap
        val objectMap = Maps.newHashMap<String, Any>()
        objectMap["key1"] = "value1"
        objectMap["key2"] = 2
        objectMap["key3"] = GeoPoint(-12.323434, 34.34534543)
        entity.objectMap = objectMap
        entity.geoPoint = GeoPoint(-12.323434, 34.34534543)
        entity.stringArray = Lists.newArrayList("a", "b")
        if (withSubcollection) {
            entity.subCollection = Lists.newArrayList(SampleEntity(), SampleEntity())
        }

        val sampleInnerEntity = SampleInnerEntity()
        sampleInnerEntity.stringField = RandomUtils.getLong().toString()
        sampleInnerEntity.longField = RandomUtils.getLong()

        entity.composite = sampleInnerEntity

        repository.add(entity)
        lastId = entity.id

        return autoMarshall(entity)
    }

    @RequestMapping(value = ["/addAll"], method = [RequestMethod.GET], produces = [MimeType.JSON_UTF8])
    @ResponseBody
    fun addAll(@RequestParam(required = false, defaultValue = "false") autoId: Boolean): String {

        val items = Lists.newArrayList<SampleEntity>()
        var entity = SampleEntity()
        if (!autoId) {
            entity.id = RandomUtils.getLong().toString()
        }
        entity.stringField = RandomUtils.getLong().toString()
        items.add(entity)
        entity = SampleEntity()
        if (!autoId) {
            entity.id = RandomUtils.getLong().toString()
        }
        entity.stringField = RandomUtils.getLong().toString()
        items.add(entity)

        repository.addAll(items)

        return autoMarshall(items)
    }

    @RequestMapping(value = ["/update"], method = [RequestMethod.GET], produces = [MimeType.JSON_UTF8])
    @ResponseBody
    fun update(): String {
        val entity = SampleEntity()
        entity.id = lastId
        entity.stringField = RandomUtils.getLong().toString()
        repository.update(entity)
        return autoMarshall(entity)
    }

    @RequestMapping(value = ["/remove"], method = [RequestMethod.GET])
    fun remove(@RequestParam id: String) {
        repository.remove(id)
    }

    @RequestMapping(value = ["/removeAll"], method = [RequestMethod.GET])
    fun removeAll() {
        repository.removeAll()
        lastId = null
    }
}